graphql_compiler/schema_merging/rename_schema.py

```diff
@@ -0,0 +1,232 @@
+
+    Args:
+        schema_string: string describing a valid schema that does not contain extensions
+        rename_func: callable that takes string to string, used to transform the names of
+                     types, interfaces, enums, and root fields. Defaults to identity
+
+    Returns:
+        RenamedSchema
+
+    Raises:
+        SchemaError if input schema_string does not represent a valid schema without extensions,
+            or if there are conflicts between the renamed types or root fields
+    """
+    # Check that the input string is a parseable and valid schema.
+    try:
+        ast = parse(schema_string)
+        build_ast_schema(ast)
+    except Exception as e:  # Can't be more specific -- see graphql/utils/build_ast_schema.py
+        raise SchemaError('Input schema does not define a valid schema.\n'
+                          'Message: {}'.format(e))
+
+    schema_data = get_schema_data(ast)
+    # Check that the input schema has no extensions
+    if schema_data.has_extension:
+        raise SchemaError('Input schema should not contain extensions.')
+
+    # Rename types, interfaces, enums
+    reverse_name_map = _rename_types(ast, rename_func, schema_data.query_type, schema_data.scalars)
+
+    # Rename root fields
+    reverse_root_field_map = _rename_root_fields(ast, rename_func, schema_data.query_type)
+
+    return RenamedSchema(schema_ast=ast, reverse_name_map=reverse_name_map,
+                         reverse_root_field_map=reverse_root_field_map)
+
+
+def _rename_types(ast, rename_func, query_type, scalars):
+    """Rename types, enums, interfaces using rename_func.
+
+    The query type will not be renamed. Scalar types, field names, enum values will not be renamed.
+
+    ast will be modified as a result.
+
+    Args:
+        ast: Document, the schema ast that we modify
+        rename_func: callable, used to rename types, interfaces, enums, etc
+        query_type: string, name of the query type, e.g. 'RootSchemaQuery'
+        scalars: set of strings, the set of user defined scalars
+
+    Returns:
+        Dict[str, str], the new type name to original type name map
+
+    Raises:
+        SchemaError if the rename causes name conflicts
+    """
+    visitor = RenameSchemaVisitor(rename_func, query_type, scalars)
+    visit(ast, visitor)
+
+    return visitor.reverse_name_map
+
+
+def _rename_root_fields(ast, rename_func, query_type):
+    """Rename root fields, aka fields of the query type.
+
+    ast will be modified as a result.
+
+    Args:
+        ast: Document, the schema ast that we modify
+        rename_func: callable, used to rename fields of the query type
+        query_type: string, name of the query type, e.g. 'RootSchemaQuery'
+
+    Returns:
+        Dict[str, str], the new root field name to original root field name map
+
+    Raises:
+        SchemaError if rename causes root field name conflicts
+    """
+    visitor = RenameRootFieldsVisitor(rename_func, query_type)
```

Comment:
RenameRootFieldsVisitor takes in 3 arguments but this gives it two. Also, this seems very similar to the function above. I am confused as to why there are not 1 function.

Reply:
The first function renames types and the second renames root fields, which feel pretty separate so I put them into two functions. The one for renaming root fields doesn't need to know about the scalar types, RenameRootFieldsVisitor takes in 2 arguments (RenameSchemaVisitor, which I will rename to something better, takes in 3 arguments). I can get rid of both of these functions and combine them into `rename_schema` if that's better?

Reply:
Nevermind I misread this.